### PR TITLE
10903 sub contact routing

### PR DESCRIPTION
**Pull request by jkuester**
All these changes have been made just with the _report_ functionality in mind.  I have not tested any support yet for contacts. :sweat: 

Also, if you are not super happy with this approach, also check out just the changes in the [first commit](https://github.com/benkags/cht-core/pull/10/commits/576b59f01076685fcd4d7a560fe8f1ba47a04f3f).  That was a more ambitious refactoring that tried to bring the `type=binary` functionality to be closer to `type=file`.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/services/enketo-translation.service.ts
Comment (jkuester) [inline webapp/src/ts/services/enketo-translation.service.ts:1]: I just reset this file to master since the original behavior of clearing the `binary` fields should work fine.

File: webapp/src/ts/services/enketo.service.ts [hunk 1]
```diff
@@ -362,17 +362,11 @@ export class EnketoService {
...
-      .each((_idx, element) => {
-        if ($(element).text() === filename) {
-          match = element;
-          return false; // break
-        }
-      });
-    return match;
+  private findFileNodeByFilename($record: JQuery<Element>, filename: string) {
+    return $record
+      .find('[type=file]')
```
Comment (jkuester) [inline webapp/src/ts/services/enketo.service.ts:367]: I think we really do need `type=file` here.  We handle the `type=binary` fields separately below.  Any field with a file being tracked in the Enketo FileManager is going to have the `type=file` (at least according to all my tests).

File: webapp/src/ts/services/enketo.service.ts [hunk 2]
```diff
@@ -478,96 +472,58 @@ export class EnketoService {
...
+      .map(element => $(element))
+      .filter($element => $element.text())
+      .forEach(($element) => {
+        const data = $element.text();
+        const { ownerDoc, ownerXpath } = resolveOwner($element[0]);
+        const xpath = Xpath
+          .getElementXPath($element[0])
+          .slice(ownerXpath.length + 1);
+        const formPrefix = ownerDoc.form ? `${ownerDoc.form}/` : '';
+        const filename = `${USER_BINARY_FILE_PREFIX}${formPrefix}${xpath}`;
```
Comment (jkuester) [inline webapp/src/ts/services/enketo.service.ts:523]: I spent an obscenely long time working on a naming approach here that uses a _hash_ of the actual base64 contents and then had the attachment name set in the doc property value (similar to the `type=file` attachments).  

It worked, but ultimately had very few benefits over this current strategy.  😓 The worst was that it ended up requiring bespoke handling code anyway. 😞 

I ended circling back to this path-based approach in the end as I think it works with the least amount of code needing to be added. (And we can just fully repurpose the "legacy" pattern for `type=binary` attachments).

File: webapp/src/ts/services/enketo.service.ts (hunk 1 above)
Comment (benkags) [reply webapp/src/ts/services/enketo.service.ts:367]: Correct. This was a bug I introduced.

Comment (benkags) [reply webapp/src/ts/services/enketo.service.ts:367]: A note: this functionality leaked from [#10922 ](https://github.com/medic/cht-core/pull/10922) where report logic was being handled but not sure if to separate out reports logic into a different PR anymore.

File: webapp/src/ts/services/enketo.service.ts (hunk 2 above)
Comment (benkags) [reply webapp/src/ts/services/enketo.service.ts:523]: I also had initially started with a `guid` as attachment name approach but that meant doing a clean up with every edit - quite inefficient.  Only note with the updated attachment name logic is we need to confirm attachment retrieval for report view also works across all edge cases: reports existing pre-this feature, sub reports view within main report and sub reports standalone view.

File: webapp/src/ts/services/format-data-record.service.ts [hunk 3]
```diff
@@ -461,21 +462,13 @@ export class FormatDataRecordService {
       return undefined;
     }
     const isImagePath = filePath => doc._attachments[filePath]?.content_type?.startsWith('image/');
-    // The field value may itself be the literal attachment name
-    // (`user-file/<xpath>`) for inline-binary fields. Required for sub-doc
-    // rendering: the parent-form prefix embedded in the name cannot be
-    // reconstructed from a sub-doc's own field path alone.
-    if (typeof value === 'string' && value.startsWith('user-file/') && isImagePath(value)) {
```
Comment (benkags) [inline webapp/src/ts/services/format-data-record.service.ts]: Just to note here that this logic here also fixed currently broken sub-reports standalone view whose attachment name was derived from the main/parent doc xpath, whose name reconstruction within the sub report fails because of missing parent doc xpath reference.

File: webapp/src/ts/services/enketo.service.ts (hunk 2 above)
Comment (jkuester) [reply webapp/src/ts/services/enketo.service.ts:523]: >  Only note with the updated attachment name logic is we need to confirm attachment retrieval for report view also works across all edge cases: reports existing pre-this feature, sub reports view within main report and sub reports standalone view.

100%!  I did test against these cases (including pulling in a report with attachment created on a `4.8.0` instance) and viewing/editing the reports worked as expected. But, it would be really good to have someone else repro this as well. :+1:

File: webapp/src/ts/services/format-data-record.service.ts (hunk 3 above)
Comment (jkuester) [reply webapp/src/ts/services/format-data-record.service.ts]: Hmm, I am not following you here.  🤔 Based on my testing (and it is totally possible I just missed something 😬) the only time I saw the _property value_ getting set to an attachment name with a corresponding `user-file/` attachment was for `type=file` attachments pre-`4.9.0`.  As far as I can tell, we have always cleared the property value for `type=binary` attachments. And, since `4.9.0` the `type=file` property values are mapped to `user-file-` attachments.
